flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/TaskExecutorTracker.java

```diff
@@ -0,0 +1,274 @@
+     *
+     * @param allocationId of the allocated slot
+     */
+    void notifyFree(AllocationID allocationId);
+
+    /**
+     * Notifies the tracker about the slot statuses.
+     *
+     * @param slotReport slot report
+     * @param instanceId of the task executor
+     */
+    void notifySlotStatus(SlotReport slotReport, InstanceID instanceId);
+
+    // ---------------------------------------------------------------------------------------------
+    // Utility method
+    // ---------------------------------------------------------------------------------------------
+
+    /**
+     * Get the instance ids of all registered task executors.
+     *
+     * @return a set of instance ids of all registered task executors.
+     */
+    Set<InstanceID> getTaskExecutors();
+
+    /**
+     * Check if there is a registered task executor with the given instance id.
+     *
+     * @param instanceId of the task executor
+     * @return whether there is a registered task executor with the given instance id
+     */
+    boolean isTaskExecutorRegistered(InstanceID instanceId);
+
+    /**
+     * Find an exactly matching pending task executor with the given resource profile.
+     *
+     * @param initialSlotReport of the task executor
+     * @param totalResourceProfile of the task executor
+     * @param defaultResourceProfile of the task executor
+     * @return An Optional of {@link PendingTaskManager}, if find, of the matching pending task
+     *     executor.
+     */
+    Optional<PendingTaskManager> findMatchingPendingTaskManager(
+            SlotReport initialSlotReport,
+            ResourceProfile totalResourceProfile,
+            ResourceProfile defaultResourceProfile);
+
+    // ---------------------------------------------------------------------------------------------
+    // TaskExecutor idleness / redundancy
+    // ---------------------------------------------------------------------------------------------
+
+    /**
+     * Get all task executors which idle exceed the given timeout period
+     *
+     * @param taskManagerTimeout timeout period
+     * @return a map of timeout task executors' connection index by the instance id
+     */
+    Map<InstanceID, TaskExecutorConnection> getTimeOutTaskExecutors(Time taskManagerTimeout);
+
+    /**
+     * Get the start time of idleness of the task executor with the given instance id.
+     *
+     * @param instanceId of the task executor
+     * @return the start time of idleness
+     */
+    long getTaskExecutorIdleSince(InstanceID instanceId);
+
+    // ---------------------------------------------------------------------------------------------
+    // slot / resource counts
+    // ---------------------------------------------------------------------------------------------
+
+    /**
+     * Get the total registered resources.
+     *
+     * @return the total registered resources
+     */
+    ResourceProfile getTotalRegisteredResources();
```

Comment:
The following methods can be aggregated to something like `StatusOverview getStatusOverview()`, where `StatusOverview` (maybe another name) is a data structure containing all information needed.
- ResourceProfile getTotalRegisteredResources()
- ResourceProfile getTotalFreeResources()
- int getNumberRegisteredSlots()
- int getNumberFreeSlots()
- int getNumberFreeTaskExecutors()
- int getNumberRegisteredTaskExecutors()
- int getNumberPendingTaskExecutors();